### PR TITLE
[Skills Everywhere][#208] Add optional script parameters to RunTestAsync

### LINKED ISSUE #208: Add optional script parameters to RunTestAsync
We have some scenarios where we would like to replace some tokens in the script before executing it (e.g.: DeliveryMode and TargetSkill are defined in the TestData parameters)

So, given the following a json script fragment:

```
    {
      "type": "message",
      "role": "user",
      "text": "skill"
    },
``` 

I should be able to change it to:

```
    {
      "type": "message",
      "role": "user",
      "text": "${TargetSkill}"
    },
```

And then execute the script passing a parameter dictionary with as follows:

```
 var scriptParams = new Dictionary<string, string>() { { "TargetSkill", testCase.TargetSkill } };
 await runner.RunTestAsync(Path.Combine(_testScriptsFolder, testCase.Script), scriptParams);
```

### PR DESCRIPTION
**Pull request by denscollo**
Fixes #208

**Note: This PR must be merged after PR #219**

## Description
This PR adds support for passing parameters from the test scripts to the TestRunner.

### Detailed Changes
- Updated **_TestRunner_** class to receive in RunTestAsync method the dictionary with the parameters to replace in the script.
- Updated **_EchoTests_** class to pass the dictionary of params with the TargetSkill to the TestRunner.
- Added EchoMultiSkill.json file with the script containing the tokens `${DeliveryMode}` and `${TargetSkill}` to be replaced.

## Testing
This image shows the EchoTests running with the changes.
image

### CODE REVIEW

File: Libraries/TranscriptTestRunner/TestRunner.cs [hunk 1]
```diff
@@ -285,8 +291,17 @@ private async Task ExecuteTestScriptAsync(string callerName, CancellationToken c
             _logger.LogInformation($"\n------ Starting test {callerName} ----------");
 
             using var reader = new StreamReader(_testScriptPath);
+            var plainTestScript = await reader.ReadToEndAsync().ConfigureAwait(false);
+
+            if (_scriptParams?.Count > 0)
```
Comment (gabog) [inline Libraries/TranscriptTestRunner/TestRunner.cs:296]: Consider replacing by a more efficient alternative (we don't need to know the exact count, we just need to know if it has elements so we can iterate over it)
```suggestion
            if (_scriptParams != null && _scriptParams.Any())
```

File: Libraries/TranscriptTestRunner/TestRunner.cs [hunk 2]
```diff
@@ -27,6 +28,7 @@ public class TestRunner
         private readonly ILogger _logger;
         private readonly int _replyTimeout;
         private readonly TestClientBase _testClient;
+        private Dictionary<string, string> _scriptParams;
```
Comment (gabog) [inline Libraries/TranscriptTestRunner/TestRunner.cs:31]: I don't think we need a field to store this, we can just take it in RunTestAsync and pass it as a parameter to ExecuteTestScriptAsync (it doesn't seem to be used again)

File: Libraries/TranscriptTestRunner/TestRunner.cs [hunk 3]
```diff
@@ -285,8 +291,17 @@ private async Task ExecuteTestScriptAsync(string callerName, CancellationToken c
             _logger.LogInformation($"\n------ Starting test {callerName} ----------");
 
             using var reader = new StreamReader(_testScriptPath);
+            var plainTestScript = await reader.ReadToEndAsync().ConfigureAwait(false);
+
+            if (_scriptParams?.Count > 0)
+            {
+                foreach (var param in _scriptParams)
+                {
+                    plainTestScript = plainTestScript.Replace("${" + param.Key + "}", param.Value);
```
Comment (gabog) [inline Libraries/TranscriptTestRunner/TestRunner.cs:300]: Can we look into using regex with multiple patterns so we traverse the script only once? (I'm not great with regex but "I think" this is doable, let me know if it isn't).

See if the following article helps: https://www.wipfli.com/insights/blogs/connect-microsoft-dynamics-365-blog/c-regex-multiple-replacements

**Review by gabog — CHANGES_REQUESTED** [review]
See comments and let me know if you have any questions.

Thanks

**Review by gabog — APPROVED** [review]

**Comment by gabog** [comment]
Thanks @denscollo , I approved the PR, feel free to merge once the conflicts are resolved.

**Comment by denscollo** [comment]
Thanks @gabog. The conflicts are solved, I'll proceed to merging the PR

**Pull request merged**